File: app/src/main/java/com/example/bulpyeonmabul/score/BoardScoreController.java
package com.example.bulpyeonmabul.score;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public final class BoardScoreController {
    private final ScoreEngine engine;
    private ScoreBoardState state;

    public BoardScoreController(List<ScorePlayerSeed> players) {
        this(players, new ScoreEngine());
    }

    public BoardScoreController(List<ScorePlayerSeed> players, ScoreEngine engine) {
        this.engine = engine;
        state = engine.createInitialState(players);
    }

    public ScoreSnapshot getSnapshot() {
        return makeSnapshot(Collections.emptyList());
    }

    public ScoreSnapshot onMoveCompleted(String playerId, int movedSpaces) {
        ScoreApplyResult result = engine.applyMoveScore(state, playerId, movedSpaces);
        state = result.getState();
        return makeSnapshot(result.getLogs());
    }

    public ScoreSnapshot onTileLanded(String playerId, BoardTileType tileType) {
        ScoreApplyResult result = engine.applyTileLanded(state, playerId, tileType);
        state = result.getState();
        return makeSnapshot(result.getLogs());
    }

    public ScoreSnapshot onAdResult(String playerId, boolean success) {
        ScoreApplyResult result = engine.applyAdResolved(state, playerId, success);
        state = result.getState();
        return makeSnapshot(result.getLogs());
    }

    public ScoreSnapshot onMiniGameFinished(List<String> rankingPlayerIds) {
        ScoreApplyResult result = engine.applyMiniGameResult(state, rankingPlayerIds);
        state = result.getState();
        return makeSnapshot(result.getLogs());
    }

    public ScoreSnapshot onChanceOrCardScoreChanged(String playerId, int delta, String reason) {
        ScoreApplyResult result = engine.applyDirectChange(state, playerId, delta, reason);
        state = result.getState();
        return makeSnapshot(result.getLogs());
    }

    public ScoreSnapshot swapScores(String firstPlayerId, String secondPlayerId, String reason) {
        ScoreApplyResult result = engine.swapScores(state, firstPlayerId, secondPlayerId, reason);
        state = result.getState();
        return makeSnapshot(result.getLogs());
    }

    public ScoreSnapshot multiplyAllScores(int multiplier, String reason) {
        ScoreApplyResult result = engine.multiplyAllScores(state, multiplier, reason);
        state = result.getState();
        return makeSnapshot(result.getLogs());
    }

    public ScoreSnapshot halveAllScoresRoundUp(String reason) {
        ScoreApplyResult result = engine.halveAllScoresRoundUp(state, reason);
        state = result.getState();
        return makeSnapshot(result.getLogs());
    }

    public ScoreSnapshot nextRound() {
        ScoreApplyResult result = engine.nextRound(state);
        state = result.getState();
        return makeSnapshot(result.getLogs());
    }

    public int getScore(String playerId) {
        for (PlayerScoreState player : state.getPlayers()) {
            if (player.getPlayerId().equals(playerId)) {
                return player.getScore();
            }
        }
        return 0;
    }

    public String getTopPlayerId() {
        List<PlayerScoreState> ranking = engine.getRanking(state);
        return ranking.isEmpty() ? null : ranking.get(0).getPlayerId();
    }

    public String getLastPlayerId() {
        List<PlayerScoreState> ranking = engine.getRanking(state);
        return ranking.isEmpty() ? null : ranking.get(ranking.size() - 1).getPlayerId();
    }

    private ScoreSnapshot makeSnapshot(List<ScoreLog> logs) {
        List<ScoreViewRow> ranking = new ArrayList<>();
        for (PlayerScoreState player : engine.getRanking(state)) {
            ranking.add(new ScoreViewRow(
                    player.getPlayerId(),
                    player.getNickname(),
                    player.getScore(),
                    player.getMiniGameWinCount()
            ));
        }
        return new ScoreSnapshot(state.getRound(), ranking, logs);
    }
}

File: app/src/main/java/com/example/adchaosdemo/score/BoardScoreController.kt
package com.example.adchaosdemo.score

data class ScoreViewRow(
    val playerId: String,
    val nickname: String,
    val score: Int,
    val miniGameWinCount: Int
)

data class ScoreSnapshot(
    val round: Int,
    val ranking: List<ScoreViewRow>,
    val logs: List<ScoreLog>
)

class BoardScoreController(
    players: List<Pair<String, String>>,
    private val engine: ScoreEngine = ScoreEngine()
) {
    private var state: ScoreBoardState = engine.createInitialState(players)

    fun getSnapshot(): ScoreSnapshot {
        return ScoreSnapshot(
            round = state.round,
            ranking = engine.getRanking(state).map { it.toViewRow() },
            logs = emptyList()
        )
    }

    fun onTileLanded(playerId: String, tileType: BoardTileType): ScoreSnapshot {
        val result = engine.applyEvent(state, ScoreEvent.TileLanded(playerId, tileType))
        state = result.state
        return makeSnapshot(result.logs)
    }

    fun onAdResult(playerId: String, success: Boolean): ScoreSnapshot {
        val result = engine.applyEvent(state, ScoreEvent.AdResolved(playerId, success))
        state = result.state
        return makeSnapshot(result.logs)
    }

    fun onMiniGameFinished(rankingPlayerIds: List<String>): ScoreSnapshot {
        val result = engine.applyEvent(state, ScoreEvent.MiniGameResult(rankingPlayerIds))
        state = result.state
        return makeSnapshot(result.logs)
    }

    fun onChanceOrCardScoreChanged(playerId: String, delta: Int, reason: String): ScoreSnapshot {
        val result = engine.applyEvent(state, ScoreEvent.DirectScoreChange(playerId, delta, reason))
        state = result.state
        return makeSnapshot(result.logs)
    }

    fun nextRound(): ScoreSnapshot {
        val result = engine.applyEvent(state, ScoreEvent.NextRound)
        state = result.state
        return makeSnapshot(result.logs)
    }

    private fun makeSnapshot(logs: List<ScoreLog>): ScoreSnapshot {
        return ScoreSnapshot(
            round = state.round,
            ranking = engine.getRanking(state).map { it.toViewRow() },
            logs = logs
        )
    }

    private fun PlayerScoreState.toViewRow(): ScoreViewRow {
        return ScoreViewRow(
            playerId = playerId,
            nickname = nickname,
            score = score,
            miniGameWinCount = miniGameWinCount
        )
    }
}

